Source/Renderer/GLSLModernizer.js

```diff
@@ -0,0 +1,233 @@
+                    splitSource[number] = line.replace(regex, replacement);
+                } else {
+                    splitSource[number] = line.replace(regex, '$1' + replacement + '$3');
+                }
+            }
+        }
+
+        function findInSource(str) {
+            var regexStr = '(^|[^\\w])(' + str + ')($|[^\\w])';
+            var regex = new RegExp(regexStr, 'g');
+            for (var number = 0; number < splitSource.length; ++number) {
+                var line = splitSource[number];
+                if (regex.test(line)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        function compileSource() {
+            var wholeSource = '';
+            for (var number = 0; number < splitSource.length; ++number) {
+                wholeSource += splitSource[number] + '\n';
+            }
+            return wholeSource;
+        }
+
+        function setAdd(variable, set1) {
+            if (set1.indexOf(variable) === -1)
+                {set1.push(variable);}
+        }
+
+        function getVariablePreprocessorBranch(variablesThatWeCareAbout) {
+            var variableMap = {};
+            var negativeMap = {};
+
+            for (var a = 0; a < variablesThatWeCareAbout.length; ++a) {
+                var variableThatWeCareAbout = variablesThatWeCareAbout[a];
+                variableMap[variableThatWeCareAbout] = [null];
+            }
+
+            var stack = [];
+            for (var i = 0; i < splitSource.length; ++i) {
+                var line = splitSource[i];
+                var hasIF = /(#ifdef|#if)/g.test(line);
+                var hasELSE = /#else/g.test(line);
+                var hasENDIF = /#endif/g.test(line);
+
+                if (hasIF) {
+                    stack.push(line);
+                } else if (hasELSE) {
+                    var top = stack[stack.length - 1];
+                    var op = top.replace('ifdef', 'ifndef');
+                    if (/if/g.test(op)) {
+                        op = op.replace(/(#if\s+)(\S*)([^]*)/, '$1!($2)$3');
+                    }
+                    stack.pop();
+                    stack.push(op);
+
+                    negativeMap[top] = op;
+                    negativeMap[op] = top;
+                } else if (hasENDIF) {
+                    stack.pop();
+                } else if (!/layout/g.test(line)) {
+                    for (var varIndex = 0; varIndex < variablesThatWeCareAbout.length; ++varIndex) {
+                        var varName = variablesThatWeCareAbout[varIndex];
+                        if (line.indexOf(varName) !== -1) {
+                            if (variableMap[varName].length === 1 && variableMap[varName][0] === null) {
+                                variableMap[varName] = stack.slice();
+                            } else {
+                                variableMap[varName] = variableMap[varName].filter(function(x) { return stack.indexOf(x) >= 0; });
```

Comment:
I'm a bit confused here. Related to `[]` comment above. Can this area be cleaner?